src/sig_stfl/lms/sig_stfl_lms_functions.c

```diff
@@ -0,0 +1,813 @@
+	 * update_private_key, context - the function that is called when the
+	 *      private key is generated; it is expected to store it to secure NVRAM
+	 *      If this is NULL, then the context pointer is reinterpretted to mean
+	 *      where in RAM the private key is expected to be placed
+	 * public_key - where to store the public key
+	 * len_public_key - length of the above buffer; see hss_get_public_key_len
+	 *      if you need a hint.
+	 * aux_data - where to store the optional aux data.  This is not required, but
+	 *      if provided, can be used to speed up the hss_generate_working_key
+	 *      process;
+	 * len_aux_data - the length of the above buffer.  This is not fixed length;
+	 *      the function will run different time/memory trade-offs based on the
+	 *      length provided
+	 *
+	 * This returns true on success, false on failure
+	 */
+	b_ret = hss_generate_private_key(
+	            LMS_randombytes,
+	            oqs_key_data->levels,
+	            oqs_key_data->lm_type,
+	            oqs_key_data->lm_ots_type,
+	            NULL, //File handler function?
+	            oqs_key_data->sec_key,
+	            oqs_key_data->public_key, len_public_key,
+	            oqs_key_data->aux_data, oqs_key_data->len_aux_data,
+	            NULL);
+	if (b_ret) {
+		memcpy(pk, oqs_key_data->public_key, len_public_key);
+		sk->secret_key_data = oqs_key_data;
+	} else {
+		OQS_MEM_secure_free(oqs_key_data->sec_key, sk->length_secret_key * sizeof(uint8_t));
+		OQS_MEM_insecure_free(oqs_key_data->aux_data);
+		OQS_MEM_insecure_free(oqs_key_data);
+		oqs_key_data = NULL;
+		return -1;
+	}
+
+	/* TODO: store key pair, file handler */
+
+	ret = 0;
+	return ret;
+}
+#endif
+
+#ifndef OQS_ALLOW_LMS_KEY_AND_SIG_GEN
+int oqs_sig_stfl_lms_sign(UNUSED OQS_SIG_STFL_SECRET_KEY *sk, UNUSED uint8_t *sm, UNUSED size_t *smlen,
+                          UNUSED const uint8_t *m, UNUSED size_t mlen) {
+	return -1;
+}
+#else
+int oqs_sig_stfl_lms_sign(OQS_SIG_STFL_SECRET_KEY *sk,
+                          uint8_t *sm, size_t *smlen,
+                          const uint8_t *m, size_t mlen) {
+
+	size_t sig_len;
+	bool status;
+	uint8_t *sig = NULL;
+	uint8_t *priv_key = NULL;
+	oqs_lms_key_data *oqs_key_data = NULL;
+	struct hss_working_key *w = NULL;
+	struct hss_sign_inc ctx;
+	if (sk) {
+		oqs_key_data = sk->secret_key_data;
+		priv_key = oqs_key_data->sec_key;
+	} else {
+		return -1;
+	}
+	w = hss_load_private_key(NULL, priv_key,
+	                         0,
+	                         NULL,
+	                         0,
+	                         0);
+	if (!w) {
+		hss_free_working_key(w);
+		return 0;
+	}
+
+	/* Now, go through the file list, and generate the signatures for each */
+
+	/* Look up the signature length */
+
+	sig_len = hss_get_signature_len_from_working_key(w);
+	if (sig_len == 0) {
+		hss_free_working_key(w);
+		return 0;
+	}
+
+	sig = malloc(sig_len);
+	if (!sig) {
+		hss_free_working_key(w);
+		return -1;
+	}
+
+	(void)hss_sign_init(
```

Comment:
Why not error-check `hss_sign_init` and `hss_sign_update`?

Reply:
For posterity: an error status is carried through in the `ctx`, and each of the incremental signing functions checks the status before doing anything. It suffices to error-check only `hss_sign_finalize`.